### PR TITLE
feat(OnyxListbox): Implement search functionality

### PR DESCRIPTION
**Pull request by JoCa96**
Relates to #706 

- Updated `OnyxListboxProps` to use discriminating unions instead of conditional generics for better type support
- Added `OnyxMiniSearch` component, which is used as part of the `OnyxListbox` to search the list
- Added `withSearch` property to `OnyxListbox` 
- renamed `ObjectToDottedStrings` to more appropriate `FlattenedKeysOf`

## Checklist

- [x] The added / edited code has been documented with [JSDoc](https://jsdoc.app/about-getting-started)
- [x] If a new component is added, at least one [Playwright screenshot test](https://github.com/SchwarzIT/onyx/actions/workflows/playwright-screenshots.yml) is added
- [x] A changeset is added with `npx changeset add` if your changes should be released as npm package (because they affect the library usage)

### CODE REVIEW

File: packages/sit-onyx/src/i18n/locales/en-US.json [hunk 1]
```diff
@@ -21,7 +21,12 @@
     "selectAll": "Select all",
     "currentSelection": "{n} selected"
   },
-  "empty": "No data available",
+  "listbox": {
+    "empty": "No data available",
+    "noMatch": "No entry matches your search",
+    "searchInputLabel": "Filter the list entries",
```
Comment (larsrickert) [inline packages/sit-onyx/src/i18n/locales/en-US.json:27]:
```suggestion
    "noMatch": "No items found matching your search",
    "searchInputLabel": "Filter the list items",
```

File: packages/sit-onyx/src/i18n/locales/de-DE.json [hunk 2]
```diff
@@ -21,7 +21,12 @@
     "selectAll": "Alle auswählen",
     "currentSelection": "{n} ausgewählt"
   },
-  "empty": "Keine Daten verfügbar",
+  "listbox": {
+    "empty": "Keine Daten verfügbar",
+    "noMatch": "Kein Element passt zum Suchtext",
+    "searchInputLabel": "Durchsuche die Liste",
+    "clearSearch": "Lösche Suchtext"
```
Comment (larsrickert) [inline packages/sit-onyx/src/i18n/locales/de-DE.json:28]:
```suggestion
    "noMatch": "Kein Element passt zum Suchtext",
    "searchInputLabel": "Liste durchsuchen",
    "clearSearch": "Suchtext löschen"
```

File: packages/sit-onyx/src/i18n/index.ts [hunk 3]
```diff
@@ -16,6 +16,8 @@ type GetTypeOfTranslations<T> = T extends object
 /** Available translations that are used by onyx components. */
 export type OnyxTranslations = GetTypeOfTranslations<typeof enUS>;
 
+export type OnyxTranslationKeys = FlattenedKeysOf<OnyxTranslations>;
```
Comment (larsrickert) [inline packages/sit-onyx/src/i18n/index.ts:19]:
```suggestion
export type OnyxTranslationKey = FlattenedKeysOf<OnyxTranslations>;
```

Its a union type, no array so I'd say its a single key type

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+</script>
+<template>
+  <div class="onyx-mini-search" v-bind="rootAttrs">
+    <input
+      v-model="value"
+      :aria-label="props.label"
+      v-bind="inputAttrs"
+      class="onyx-mini-search__input"
+      placeholder="Search"
+      type="text"
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:50]:
```suggestion
      type="search"
```

Type search also already has a clear icon that we can maybe style more easily and we might get right of the rotoAttrs and inputAttrs above

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    />
+    <!-- We use `@mousedown.prevent` here to not lose the input focus when the button is clicked  -->
+    <button
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:53]: button should be hidden when the search value is empty, right?

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import { computed, useAttrs, type HtmlHTMLAttributes } from "vue";
+import OnyxIcon from "../OnyxIcon/OnyxIcon.vue";
+import xSmall from "@sit-onyx/icons/x-small.svg?raw";
+import { injectI18n } from "../../i18n";
+
+defineOptions({ inheritAttrs: false });
+
+const attrs = useAttrs();
+
+const props = defineProps<{ modelValue: string; label: string }>();
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:11]: Should we align with other components and create a `types.ts` file with the props?

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:1]: Minisearch is a fulltext search library, should we maybe rename this component so something like `OnyxListboxSearch`?:
https://github.com/lucaong/minisearch

File: packages/sit-onyx/src/components/OnyxListbox/OnyxListbox.vue [hunk 7]
```diff
@@ -130,7 +140,12 @@ const { vScrollEnd, isScrollEnd } = useScrollEnd({
   offset: computed(() => props.lazyLoading?.scrollOffset),
 });
 
-const isEmpty = computed(() => props.options.length === 0);
+const isEmptyMessage = computed<FlattenedKeysOf<OnyxTranslations> | undefined>(
+  () =>
+    (props.options.length === 0 && "listbox.empty") ||
+    (props.withSearch && props.searchTerm && props.options.length === 0 && "listbox.noMatch") ||
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxListbox/OnyxListbox.vue:146]: listbox.noMatch can never be reached because it checks for `props.options.length === 0` which is always returned earlier with "listbox.empty". Also imho I think the `&&` syntax is a bit hard to understand. What do you think about this?

```ts
const isEmptyMessage = computed(() => {
  if (props.options.length === 0) return t.value("listbox.empty");
  if (props.withSearch && props.searchTerm && props.options.length === 0) {
    return t.value("listbox.noMatch");
  }
  return undefined;
});
```

Comment (larsrickert) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxListbox.vue:146]: Also I'd ususally prefer to directly return the translated message instead of the key

File: packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png
Comment (larsrickert) [inline packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png:1]: Would prefer to use our matrix screenshot util and also add test cases for the empty states (no options and no search matches)

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue (hunk 4 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:50]: The clear icon is User-Agent dependent and non-standard, see also: https://developer.mozilla.org/en-US/docs/Web/CSS/::-webkit-search-cancel-button
Therefore I would like to stay with type `text`

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:1]: The name is aligned with Figma, so I would prefer to keep the name.

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue (hunk 5 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:53]: sounds good

File: packages/sit-onyx/src/i18n/locales/en-US.json (hunk 1 above)
Comment (JoCa96) [reply packages/sit-onyx/src/i18n/locales/en-US.json]: updated

File: packages/sit-onyx/src/i18n/index.ts (hunk 3 above)
Comment (JoCa96) [reply packages/sit-onyx/src/i18n/index.ts]: sounds good

File: packages/sit-onyx/src/components/OnyxListbox/OnyxListbox.vue (hunk 7 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxListbox.vue]: fixed

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue (hunk 6 above)
Comment (larsrickert) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:11]: Then I'd at least move it to a dedicated type and export it from the .vue file so it can be easily used :)

File: packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png
Comment (JoCa96) [reply packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png:1]: done

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue (hunk 6 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue]: done

File: packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png
Comment (JoCa96) [reply packages/sit-onyx/playwright/snapshots/components/OnyxListbox/with-search-empty-chromium-linux.png:1]: done

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    <input
+      v-model="value"
+      :aria-label="props.label"
+      v-bind="inputAttrs"
+      class="onyx-mini-search__input"
+      placeholder="Search"
+      type="text"
+    />
+    <!-- We use `@mousedown.prevent` here to not lose the input focus when the button is clicked  -->
+    <button
```
[the remark this answers is not in the document]
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:55]: the width gets jumpy if we do that

File: packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      "class" | "style"
+    >,
+);
+
+const inputAttrs = computed<Omit<HtmlHTMLAttributes, "class" | "style">>(() => {
+  const rest = { ...attrs };
+  delete rest.class;
+  delete rest.style;
+  return rest;
+});
```
[the remark this answers is not in the document]
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxListbox/OnyxMiniSearch.vue:42]: Want to avoid a wrapper, so all attributes that effect style are applied to the root, everything else applies to the input.

**Review by larsrickert — APPROVED** [review]
